Clean-up find_symbols

romainbrenguier opened the pull request:
This adds a couple of functions which are a bit more practical to use, deprecate one, rename some functions which are ambiguous and document several of these functions.

 
<!---
Thank you for your contribution. Please make sure your pull request fulfils all of the below requirements. If you cannot currently tick all the boxes, but would still like to create a PR, then add the label "work in progress" and assign the PR to yourself.
--->

- [x] Each commit message has a non-empty body, explaining why the change was made.
- [ ] Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- [na] The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [ ] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- [na] My commit message includes data points confirming performance improvements (if claimed).
- [x] My PR is restricted to a single feature or bugfix.
- [na] White-space or formatting changes outside the feature-related changed lines are in commits of their own.

<!---
See, e.g., https://chris.beams.io/posts/git-commit/ for general guidelines on commit messages.

If you have created commits mixing multiple features and/or unrelated white-space changes, use a sequence involving git reset and git add -p to fix this.
--->

romainbrenguier committed:
Add a find_symbols function returning a value

This is more intuitive to use than one modifying its argument.

romainbrenguier committed:
Use set returning find_symbols

Simpler to use

romainbrenguier committed:
Use set<symbol_exprt> version of find_symbols

This avoids having to then filter on ID_symbol

romainbrenguier committed:
Use set returning version of find_symbols

This is simpler to use.

romainbrenguier committed:
Use set<symbol_exprt> version of find_symbols

Avoid having to then filter on ID_symbol

romainbrenguier committed:
Use set returning version of find_symbols

This is simpler to use

romainbrenguier committed:
Deprecate set<exprt> version of find_symbols

This is unused

romainbrenguier committed:
Define a find_symbols taking iterators

This will make it easier to accumulate the set of symbols from a range,
which is often how the find_symbols function is used.

romainbrenguier committed:
Simplify function using find_symbols on iterators

The code is much simpler and easy to understand.

romainbrenguier committed:
Document find_symbols functions

The difference between some versions is not obvious

romainbrenguier committed:
Simplify using iterator version of find_symbols

This allows to make more things const and removes a loop, so the code is
easier to reason about.

romainbrenguier committed:
Rename function to find_symbols_or_nexts

This avoids ambiguity with the other find_symbols functions

smowton commented on src/goto-instrument/concurrency.cpp:
```diff
@@ -143,42 +143,38 @@ void concurrency_instrumentationt::instrument(
 
 void concurrency_instrumentationt::collect(const exprt &expr)
```
Commit message: typo "fild"

allredj reviewed (commented):
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: 98008c15686d5b5cabb190b79fb54280b7f4e923).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/115833500

peterschrammel commented on src/util/find_symbols.h:
```diff
@@ -39,6 +39,8 @@ void find_symbols(
   const exprt &src,
   std::set<symbol_exprt> &dest);
 
+std::set<symbol_exprt> find_symbols(const exprt &src);
```
add doxygen

romainbrenguier replied:
Done in the "Document find_symbols functions" commit

peterschrammel commented on src/goto-instrument/concurrency.cpp:
```diff
@@ -143,42 +143,38 @@ void concurrency_instrumentationt::instrument(
   find_symbols(expr, symbols);
 
-  for(std::set<exprt>::const_iterator
+  for(std::set<symbol_exprt>::const_iterator
```
ranged for?

romainbrenguier replied:
done in a later commit

peterschrammel commented on src/goto-instrument/concurrency.cpp:
```diff
@@ -143,16 +143,9 @@ void concurrency_instrumentationt::instrument(
+  for(const auto &symbol_expr : find_symbols(expr))
   {
-    const irep_idt identifier = to_symbol_expr(*s_it).get_identifier();
+    const irep_idt identifier = symbol_expr.get_identifier();
```
const &

peterschrammel commented on src/goto-instrument/concurrency.cpp:
```diff
@@ -78,34 +78,31 @@ class concurrency_instrumentationt
   replace_symbolt replace_symbol;
 
-  for(std::set<exprt>::const_iterator
+  for(std::set<symbol_exprt>::const_iterator
```
ranged for?

peterschrammel commented on src/goto-instrument/accelerate/acceleration_utils.cpp:
```diff
@@ -211,23 +211,14 @@ void acceleration_utilst::stash_variables(
-
-  irep_idt loop_counter_name=to_symbol_expr(loop_counter).get_identifier();
+  find_symbols_sett vars = find_symbols(modified.begin(), modified.end());
+  const irep_idt loop_counter_name =
```
const &

peterschrammel commented on src/util/find_symbols.h:
```diff
@@ -26,6 +27,17 @@ void find_symbols(
+    std::is_base_of<exprt, typename iteratort::value_type>::value,
+    "find_symbols takes exprt iterators as arguments");
+  find_symbols_sett result;
+  std::for_each(begin, end, [&](const exprt &e){find_symbols(e, result);});
```
little whitespace?

allredj reviewed (commented):
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: 57897ca1b63ea13b1f461efa91fb3724b3793f99).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/115947257

tautschnig commented on src/goto-instrument/concurrency.cpp:
```diff
@@ -143,16 +143,9 @@ void concurrency_instrumentationt::instrument(
+  for(const auto &symbol_expr : find_symbols(expr))
   {
-    const irep_idt identifier = to_symbol_expr(*s_it).get_identifier();
+    const irep_idt &identifier = symbol_expr.get_identifier();
```
Isn't there also a version where we just collect the identifiers rather than the full expressions?

romainbrenguier replied:
No exactly. There is `find_symbols(const exprt &, find_symbols_sett &, bool, bool)` but it wouldn't be practical to use it here. I'm adding a `find_symbol_identifiers` function to be use there, that should be better.

tautschnig commented on src/ansi-c/expr2c.cpp:
```diff
@@ -113,7 +113,7 @@ static std::string clean_identifier(const irep_idt &id)
 {
   find_symbols_sett symbols;
-  find_symbols(expr, symbols);
+  find_symbols_or_nexts(expr, symbols);
```
Yes and no: ok, this is what the function does, but it's not actually what we want. Or at least not in all places. "next" symbols should be handled transparently in verification back-ends, but language front-ends have no business dealing with them.

romainbrenguier committed:
Define a find_symbol_identifiers function

This compute a similar set to the find_symbols(const exprt &,
find_symbols_sett &, bool, bool) function but returns the computed set.

romainbrenguier committed:
Use find_symbol_identifiers in expr2c

Instead of find_symbol_or_nexts. Language front-end should not have to
deal with "next symbols".

allredj reviewed (commented):
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: 3abe34e628259e41aa89819c5b60ca90b9660edc).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/116107681
